pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,375 @@
+		candidates[id] = c()
+	}
+
+	return &Store{
+		store:       make(map[Kind]map[string]Entity),
+		subscribers: []subscriber{},
+
+		candidates: candidates,
+		collectors: make(map[string]Collector),
+		eventCh:    make(chan []Event, eventChBufferSize),
+	}
+}
+
+// Run starts the workload metadata store.
+func (s *Store) Run(ctx context.Context) {
+	retryTicker := time.NewTicker(retryCollectorInterval)
+	pullTicker := time.NewTicker(pullCollectorInterval)
+	health := health.RegisterLiveness("workloadmeta-store")
+
+	// Start collectors immediately
+	s.startCandidates(ctx)
+
+	// Start a pull immediately to fill the store without waiting for the
+	// next tick.
+	pullCtx, pullCancel := context.WithTimeout(ctx, pullCollectorInterval)
+	s.pull(pullCtx)
+
+	log.Info("workloadmeta store initialized successfully")
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+
+			case <-pullTicker.C:
+				// pullCtx will always be expired at this point
+				// if pullTicker has the same duration as
+				// pullCtx, so we cancel just as good practice
+				pullCancel()
+
+				pullCtx, pullCancel = context.WithTimeout(ctx, pullCollectorInterval)
+				s.pull(pullCtx)
+
+			case evs := <-s.eventCh:
+				s.handleEvents(evs)
+
+			case <-retryTicker.C:
+				s.startCandidates(ctx)
+
+				if len(s.candidates) == 0 {
+					retryTicker.Stop()
+				}
+
+			case <-ctx.Done():
+				retryTicker.Stop()
+				pullTicker.Stop()
+
+				err := health.Deregister()
+				if err != nil {
+					log.Warnf("error de-registering health check: %s", err)
+				}
+
+				return
+			}
+		}
+	}()
+}
+
+// Subscribe returns a channel where workload metadata events will be streamed
+// as they happen.
+func (s *Store) Subscribe(name string, filter *Filter) chan EventBundle {
+	// ch needs to be buffered since we'll send it events before the
+	// subscriber has the chance to start receiving from it. if it's
+	// unbuffered, it'll deadlock.
+	sub := subscriber{
+		name:   name,
+		ch:     make(chan EventBundle, 1),
+		filter: filter,
+	}
+
+	s.subscribersMut.Lock()
+	s.subscribers = append(s.subscribers, sub)
+	s.subscribersMut.Unlock()
+
+	var events []Event
+
+	s.storeMut.RLock()
+	if len(s.store) > 0 {
+		for kind, entitiesOfKind := range s.store {
+			if !sub.filter.MatchKind(kind) {
+				continue
+			}
+
+			// TODO(juliogreff): implement filtering by source once
+			// each source has its own separate store. since at the
+			// time of writing there's a single source, this should
+			// not matter.
+
+			for _, entity := range entitiesOfKind {
+				events = append(events, Event{
+					// TODO(juliogreff): insert Source here
+					// after the above TODO has been
+					// addressed.
+					Type:   EventTypeSet,
+					Entity: entity,
+				})
+			}
+		}
+	}
+	s.storeMut.RUnlock()
+
+	// notifyChannel should not wait when doing the first subscription, as
+	// the subscriber is not ready to receive events yet
+	notifyChannel(sub.name, sub.ch, events, false)
```

Comment:
Maybe another way to avoid this deadlock is to perform this notification in a goroutine?

Reply:
Starting a goroutine implies knowing when it will stop. I find that it requires less coordination to just not have blocking behavior to begin with. As discussed elsewhere in the thread here, we may also add an option to subscribers opt out of blocking behavior entirely, so this may become more useful in the future too.

Reply:
That makes sense.  In this case, the runtime of notifyChannel is bounded anyway, so my thinking was that you do know when it will stop!